[infra] nightly pypi build for `pyiceberg_core`

kevinjqliu opened the pull request:
Inspired by [pyiceberg#1601](https://github.com/apache/iceberg-python/pull/1601) & https://github.com/apache/iceberg-python/blob/main/.github/workflows/nightly-pypi-build.yml

This PR adds nightly build of `pyiceberg_core` to testpypi. 

I have successfully tested the github action on my fork by pushing to my own package,
* Action run: https://github.com/kevinjqliu/iceberg-rust/actions/runs/13381378488
* testpypi artifact: https://test.pypi.org/project/pyiceberg-core-kevinliu/

I've also verified library versions locally
```
(3.12.8) ➜  python git:(kevinjqliu/nightly-pyiceberg-core) pip install -i https://test.pypi.org/simple/ pyiceberg-core-kevinliu --prefer-binary --force
Looking in indexes: https://test.pypi.org/simple/
Collecting pyiceberg-core-kevinliu
  Using cached https://test-files.pythonhosted.org/packages/4b/c7/771857185d3d96e4ce0b59293540d0e1df9a8a39dbc938e0770328504c25/pyiceberg_core_kevinliu-0.4.0.dev20250218012743-cp39-abi3-macosx_10_12_x86_64.macosx_11_0_arm64.macosx_10_12_universal2.whl.metadata (1.7 kB)
Using cached https://test-files.pythonhosted.org/packages/4b/c7/771857185d3d96e4ce0b59293540d0e1df9a8a39dbc938e0770328504c25/pyiceberg_core_kevinliu-0.4.0.dev20250218012743-cp39-abi3-macosx_10_12_x86_64.macosx_11_0_arm64.macosx_10_12_universal2.whl (1.3 MB)
Installing collected packages: pyiceberg-core-kevinliu
  Attempting uninstall: pyiceberg-core-kevinliu
    Found existing installation: pyiceberg_core-kevinliu 0.4.0.dev20250218012743
    Uninstalling pyiceberg_core-kevinliu-0.4.0.dev20250218012743:
      Successfully uninstalled pyiceberg_core-kevinliu-0.4.0.dev20250218012743
Successfully installed pyiceberg-core-kevinliu-0.4.0.dev20250218012743

pip show pyiceberg-core-kevinliu

(3.12.8) ➜  python git:(kevinjqliu/nightly-pyiceberg-core) pip show pyiceberg-core-kevinliu
Name: pyiceberg_core-kevinliu
Version: 0.4.0.dev20250218012743
Summary: 
Home-page: https://rust.iceberg.apache.org
Author: 
Author-email: 
License: Apache-2.0
Location: /Users/kevinliu/.pyenv/versions/3.12.8/lib/python3.12/site-packages
Requires: 
Required-by: 
```

kevinjqliu commented on .github/workflows/release_python_nightly.yml:
```diff
@@ -0,0 +1,111 @@
+  contents: read
+
+jobs:
+  sdist:
```
TODO: add gate on apache/iceberg-rust repo

kevinjqliu commented on .github/workflows/release_python_nightly.yml:
```diff
@@ -0,0 +1,111 @@
+    # - name: Publish to TestPyPI
+    #   uses: pypa/gh-action-pypi-publish@release/v1
+    #   with:
+    #     repository-url: https://test.pypi.org/legacy/
```
TODO: setup trusted publisher for https://test.pypi.org/project/pyiceberg-core/

kevinjqliu commented on .github/workflows/release_python_nightly.yml:
```diff
@@ -0,0 +1,113 @@
+
+    environment:
+      name: testpypi
+      url: https://test.pypi.org/p/pyiceberg_core
```
TODO: perhaps call this `pyiceberg_core_nightly`?

Xuanwo replied:
Makes sense to me. Please make sure this package is owned by Iceberg PMC members.

Fokko replied:
For consistency I think `pyiceberg_core` is more in line with `pyiceberg`, right? We don't host `pyiceberg_nightly` on `test.pypi.org`. Am I missing something?

kevinjqliu replied:
I like `pyiceberg_core` for consistency too. 
For `pyiceberg`, because of the version naming convention, we had to yank current releases to avoid accidentally installing the wrong version. We'd want to do the same for `pyiceberg_core`
https://test.pypi.org/project/pyiceberg-core/

Fokko replied:
@sungwy Can you give us access? :)

Xuanwo commented on .github/workflows/release_python_nightly.yml:
```diff
@@ -0,0 +1,113 @@
+            args: --release -o dist
+          env:
+            # Workaround ring 0.17 build issue
+            CFLAGS_aarch64_unknown_linux_gnu: "-D__ARM_ARCH=8"
```
This workaround is wrong and need to fix. Please refer to https://github.com/apache/opendal/pull/5522 for changes, and https://github.com/apache/opendal/issues/5483 for details.

kevinjqliu replied:
I copied this over from the original `release_python.yml`
https://github.com/apache/iceberg-rust/blob/562ec947df35d172306c078fc42777f6d10b060f/.github/workflows/release_python.yml#L84-L85

Let me open a fix separately and then rebase on that change

kevinjqliu replied:
#975

kevinjqliu replied:
also applied the same change to this PR

kevinjqliu commented:
Rebased and addressed @Xuanwo 's concern with building wheels. 

I'll set up trusted publishing for `pyiceberg_core` in testpypi when I get the proper permission :)

Xuanwo reviewed (approved):
Looks great to me. Let see how @Fokko and @liurenjie1024 think about this.

kevinjqliu commented:
Added trusted published to testpypi's `pyiceberg-core` (Thanks sung)

![Screenshot 2025-02-20 at 8 33 11 AM](https://github.com/user-attachments/assets/ebfbb09e-2e86-47a9-9dc3-05a0dbd4b239)
